cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,391 @@
+	rootArgs := &rootArgs{}
+	cmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "Upgrade Istio control plane in-place",
+		Long: "The mesh upgrade command checks for upgrade version eligibility and," +
+			" if eligible, upgrades the Istio control plane components in-place. Warning: " +
+			"traffic may be disrupted during upgrade. Please ensure PodDisruptionBudgets " +
+			"are defined to maintain service continuity.",
+		Example: `mesh upgrade`,
+		RunE: func(cmd *cobra.Command, args []string) (e error) {
+			l := newLogger(rootArgs.logToStdErr, cmd.OutOrStdout(), cmd.OutOrStderr())
+			return upgrade(rootArgs, macArgs, l)
+		},
+	}
+	addFlags(cmd, rootArgs)
+	addUpgradeFlags(cmd, macArgs)
+	return cmd
+}
+
+// upgrade is the main function for Upgrade command
+func upgrade(rootArgs *rootArgs, args *upgradeArgs, l *logger) (err error) {
+	initLogsOrExit(rootArgs)
+	l.logAndPrintf("Client - istioctl version: %s\n", opversion.OperatorVersionString)
+
+	// Generates values for args.inFilename ICP specs yaml
+	targetValues, err := genProfile(true, args.inFilename, "",
+		"", "", args.force, l)
+	if err != nil {
+		l.logAndPrintf("Abort. Failed to generate values from file: %v, error: %v", args.inFilename, err)
+		return err
+	}
+
+	// Generate ICPS objects
+	_, targetICPS, err := genICPS(args.inFilename, "", "", args.force, l)
+	if err != nil {
+		l.logAndPrintf("Failed to generate ICPS from file %s, error: %s",
+			args.inFilename, err)
+		return err
+	}
+
+	// Get the target version from the tag in the ICPS
+	targetVersion := targetICPS.GetTag()
+	l.logAndPrintf("Upgrade - target version: %s\n", targetVersion)
+
+	// Create a kube client from args.kubeConfigPath and  args.context
+	kubeClient := getKubeExecClient(args.kubeConfigPath, args.context, l)
+
+	// Get Istio control plane namespace
+	//TODO(elfinhe): support components distributed in multiple namespaces
+	istioNamespace := targetICPS.GetDefaultNamespace()
+
+	// Read the current Istio version from the the cluster
+	currentVer, err := retrieveControlPlaneVersion(kubeClient, istioNamespace, l)
+	if err != nil {
+		l.logAndPrintf("Failed to read the current Istio version, error: %v", err)
+		return err
+	}
+
+	// Read the current Istio installation values from the cluster
+	currentValues, err := readValuesFromInjectorConfigMap(kubeClient, istioNamespace)
+	if err != nil {
+		l.logAndPrintf("Failed to read the current Istio installation values, "+
+			"error: %v", err)
+		return err
+	}
+
+	if !args.force {
+		// Check if the upgrade currentVer -> targetVersion is supported
+		err := checkSupportedVersions(currentVer, targetVersion, args.versionsURI, l)
+		if err != nil {
+			l.logAndPrintf("Upgrade version check failed: %v -> %v. Error: %v",
+				currentVer, targetVersion, err)
+			return err
+		}
+		l.logAndPrintf("Upgrade version check passed: %v -> %v.\n", currentVer, targetVersion)
+
+		checkUpgradeValues(currentValues, targetValues, l)
+		waitForConfirmation(args.yes, l)
```

Comment:
i think waitForConfirmation should be outside this block. even when forcing, yes is independent of it.